text/0171-cloudfront-redesign.md

```diff
@@ -0,0 +1,484 @@
+name, you must associate a certificate with your distribution that contains your domain name. The certificate must be present in the AWS Certificate
+Manager (ACM) service in the US East (N. Virginia) region; the certificate may either be created by ACM, or created elsewhere and imported into ACM.
+
+```ts
+const myCertificate = new certmgr.DnsValidatedCertificate(this, 'mySiteCert', {
+  domainName: 'www.example.com',
+  hostedZone,
+});
+new cloudfront.Distribution(this, 'myDist', {
+  origin: cloudfront.Origin.fromBucket(myBucket),
+  certificate: myCertificate,
+});
+```
+
+Note that in the above example the aliases are inferred from the certificate and do not need to be explicitly provided.
+
+## Caching Behaviors
+
+Each distribution has a default cache behavior which applies to all requests to that distribution; additional cache behaviors may be specified for a
+given URL path pattern. Cache behaviors allowing routing with multiple origins, controlling which HTTP methods to support, whether to require users to
+use HTTPS, and what query strings or cookies to forward to your origin, among other behaviors.
+
+The properties of the default cache behavior can be adjusted as part of the distribution creation. The following example shows configuring the HTTP
+methods and viewer protocol policy of the cache.
+
+```ts
+const myWebDistribution = new cloudfront.Distribution(this, 'myDist', {
+  origin: cloudfront.Origin.fromHTTPServer({
+    domainName: 'www.example.com',
+    originProtocolPolicy: OriginProtocolPolicy.HTTPS_ONLY,
+  }),
+  behavior: {
+    allowedMethods: AllowedMethods.ALL,
+    viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
+  }
+});
+```
+
+Additional cache behaviors can be specified at creation, or added to the origin(s) after the initial creation. These additional cache behaviors enable
+customization for a specific set of resources based on a URL path pattern. For example, we can add a behavior to `myWebDistribution` to override the
+default time-to-live (TTL) for all of the images.
+
+```ts
+myWebDistribution.origin.addBehavior('/images/*.jpg', {
+  viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
+  defaultTtl: cdk.Duration.days(7),
+});
+```
+
+## Multiple Origins
+
+A distribution may have multiple origins in addition to the default origin; each additional origin must have (at least) one behavior to route requests
+to that origin. A common pattern might be to serve all static assets from an S3 bucket, but all dynamic content served from a web server. The
+following example shows how such a setup might be created:
+
+```ts
+const myWebsiteBucket = new s3.Bucket(...);
+const myMultiOriginDistribution = new cloudfront.Distribution(this, 'myDist', {
+  origin: cloudfront.Origin.fromWebsiteBucket(myBucket),
+  additionalOrigins: [
+    cloudfront.Origin.fromLoadBalancerV2(myLoadBalancer, {
+      pathPattern: '/api/*',
```

Comment:
I think rather than double down on the distinction between `Origin` and `AdditionalOrigin` (which is just going to explode cases), can we do the reverse? Treat all origins syntactically the same:

```
origins: Origin[]
```

But do a runtime check to make sure the array contains exactly one `Origin` without a `pathPattern`?

Reply:
@rix0rrr  - I actually started out with the flat list of origins, but @eladb pointed out that if 90% of the time the user only configured a single origin, we should make that friendlier and not require an array.

Reply:
Yes, we usually try to optimize for the common use case. I like the current design where the primary original is modeled as `origin` and is required.  and also has some other requirements that can be enforced at compile time.

Another way to look at it is that in practice CloudFront itself has a concept of the primary origin (whether semantically defines or not)

Reply:
It might also be worth renaming ‘origin’ to ‘primaryOrigin’